lib/src/repo.rs

```diff
@@ -1664,34 +1665,27 @@ impl MutableRepo {
     }
 
     fn merge_view(&mut self, base: &View, other: &View) -> BackendResult<()> {
-        // TODO: Use `diff_named_commit_ids` to simplify this.
-        // Merge working-copy commits. If there's a conflict, we keep the self side.
-        for (workspace_id, base_wc_commit) in base.wc_commit_ids() {
-            let self_wc_commit = self.view().get_wc_commit_id(workspace_id);
-            let other_wc_commit = other.get_wc_commit_id(workspace_id);
-            if other_wc_commit == Some(base_wc_commit) || other_wc_commit == self_wc_commit {
-                // The other side didn't change or both sides changed in the
-                // same way.
-            } else if let Some(other_wc_commit) = other_wc_commit {
-                if self_wc_commit == Some(base_wc_commit) {
+        // In cases when base_target would equal to other_target, or neither would
+        // exist, the conflict resolves to the `self` position and there is nothing we
+        // need to do. So, it's fine that `diff_named_commit_ids`` does not return such
+        // pairs
+        let changed_wc_commits = diff_named_commit_ids(base.wc_commit_ids(), other.wc_commit_ids());
+        for (workspace_id, (base_target, other_target)) in changed_wc_commits {
+            let self_target = self.view().get_wc_commit_id(workspace_id);
+            if let Some(other_target) = other_target {
+                if base_target == self_target {
                     self.view_mut()
-                        .set_wc_commit(workspace_id.clone(), other_wc_commit.clone());
+                        .set_wc_commit(workspace_id.clone(), other_target.clone());
+                } else {
+                    // If there's a conflict, we keep the `self` side unchanged
                 }
             } else {
                 // The other side removed the workspace. We want to remove it even if the self
-                // side changed the working-copy commit.
+                // side changed the working-copy commit. TODO(ilyagr): I'm not sure why
```

Comment:
Continuing the discussion from https://github.com/jj-vcs/jj/pull/5901#discussion_r1990499773, why do you think this makes sense?

I itch to "fix" this (not in this PR), but I'm not sure what the consequences would be.

Reply:
Quoting @yuja from #5901:

> I assume the expectation is that jj workspace forget is stronger than e.g. jj new because the workspace is forgotten anyway. I don't know if that's always true, but it would be more surprising if jj workspace forget was cancelled?

Reply:
I guess `jj workspace forget` is the only case a workspace disappears? If so, it makes sense that, whether or not this is a bug, people would encounter it very rarely.